README.md
# Menu de aplicaciones basado en micropython y placa NodeMCU ESP8266

Para este ejemplo se esta usando una placa NodeMCU ESP8266 con firmware de micropython versión de firmware
**<a href="https://micropython.org/resources/firmware/esp8266-20190529-v1.11.bin">esp8266-20190529-v1.11.bin</a>**
el cual hs sido cargado siguiendo las 
**<a href="http://docs.micropython.org/en/latest/esp8266/tutorial/intro.html#deploying-the-firmware">instrucciones</a>**
del sitio de micropython.
Se esta usando los drivers para chipset **<a href="https://www.silabs.com/products/development-tools/software/usb-to-uart-bridge-vcp-drivers">CP2102</a>**.

También es necesario contar con la aplicación de consola
**<a href="https://learn.adafruit.com/micropython-basics-load-files-and-run-code/install-ampy">ampy de Adafruit</a>**
 para poder cargar los scripts de python a la placa NodeMCU

A continuación la tabla de conexiones:

| NodeMCU | Rotary Encoder |
| ------:| -----------:|
| D0 | Push Button (PB) |
| D3 | PinA |
| D4 | PinB |
| 3.3 | PullUp PinA/PinB/PB |
| GND | Middle Pin/PB |

Versión SPI

| NodeMCU | Display SPI 128x64 |
| ------:| -----------:|
| D1 | RES |
| D2 | DC |
| 3.3 | VDD |
| GND | GND |
| D5 | SCK |
| D7 | SDA |
| D8 | CS |

Versión I2C

| NodeMCU | Display I2C 128x64 |
| ------:| -----------:|
| 3.3 | VDD |
| GND | GND |
| D5 | SCK |
| D7 | SDA |


En la figura 1 se muestra de forma de conectar los componentes, para el caso de I2C se deben conectar solo los pines del 1 al 4, el resto se omite.

<center> <img src=img01.png> <br/>Figura 1 </center>

Se deben de copiar todos los archivos que se encuentren en la carpeta **simple_menu** hacia la raíz de la placa de la siguiente forma:

~~~bash
ampy -p com3 put main.py
ampy -p com3 put ssd1306.py
ampy -p com3 put simple_encoder.py
ampy -p com3 put simple_menu.py
ampy -p com3 put clock.py
ampy -p com3 put reset.py
ampy -p com3 put settings.json
~~~

com3 es el puerto Windows donde ha quedado instalado el controlador del chip usb-serial CH340G o CP2102 dependiendo del modelo de la placa. Si se tiene Linux el puerto generalmente es /dev/ttyUSB0


El archivo main.py es el primer programa que se carga al inicializar o reiniciar la placa. Aquí se muestra cu contenido en donde se puede ver como se inicializa el objeto *display* que se reutiliza en los módulos cargados, también se muestra la creación del menú y sus acciones de cada elemento del menú, dentro de cada aplicación puede contener un menú (submenu) o un programa final, dicho programa final debe tener una forma de finalizar para regresar el control al menú de donde se ejecutó. Se recomienda que la terminación del programa sea pulsando el push button del rotary encoder. Como ejemplo existe el archivo clock.py
~~~micropython
import time
import ssd1306
import simple_menu
from machine import SPI, Pin
# from machine import I2C, Pin

pb = Pin(16, Pin.IN)

# i2c = I2C(-1, Pin(14), Pin(13))
# display = ssd1306.SSD1306_I2C(128, 64, i2c, addr=0x3C, external_vcc=False)
display = ssd1306.SSD1306_SPI(128,64,SPI(1), Pin(4), Pin(5), Pin(15), False)

menu_list = ['back', 'clock', 'alarm', 'wifi connect', 'settings', 'reset']
menu_actions = ['<BACK>', 'clock', 'alarm', 'wificonnect', 'settings', 'reset']
menu = simple_menu.simple_menu(display, pb, menu_list, menu_actions)

while True:
    display.fill_rect(0,0,128,64,0)
    display.text("{: ^16s}".format("WELCOME"),0,0,1)
    display.text("{: ^16s}".format("PRESS START"),0,30,1)
    display.show()

    if not pb.value():
        menu.show()
    time.sleep_ms(10)
~~~

Algunos videos de Youtube para complementar información:

<a href="https://www.youtube.com/channel/UCOhshmzwIlELoF38J2j9k-w">Canal de Youtube</a>



CHANGES.md
# CHANGELOG

## 0.0.1 (2020-01-02)
- Carga inicial

simple_menu/clock.py
import time
import ssd1306
from machine import Pin, RTC
import math

class clock():
    def __init__(self, display, rotary_enc, settings_obj):
        self.display = display
        self.rotary_enc = rotary_enc
        self.settings_obj = settings_obj

        led = Pin(2, Pin.OUT)
        led.on()

        self.display.fill(0)
        self.display.text("",0,0,1)
        self.display.text("",0,12,1)
        self.display.text("",0,24,1)
        self.display.text("",0,36,1)
        self.display.text("",0,48,1)

        ICON = [
            [ 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [ 0, 1, 1, 0, 0, 0, 1, 1, 0],
            [ 1, 1, 1, 1, 0, 1, 1, 1, 1],
            [ 1, 1, 1, 1, 1, 1, 1, 1, 1],
            [ 1, 1, 1, 1, 1, 1, 1, 1, 1],
            [ 0, 1, 1, 1, 1, 1, 1, 1, 0],
            [ 0, 0, 1, 1, 1, 1, 1, 0, 0],
            [ 0, 0, 0, 1, 1, 1, 0, 0, 0],
            [ 0, 0, 0, 0, 1, 0, 0, 0, 0],
        ]

        for y, row in enumerate(ICON):
            for x, c in enumerate(row):
                self.display.pixel(x+118, y+0, c)
        self.display.show()

        iterations = 210
        radio = 32
        for i in range(iterations):
            my_sin = math.sin(i*2*math.pi/iterations)
            my_cos = math.cos(i*2*math.pi/iterations)
            self.display.pixel(96+int(radio*my_cos),32+int(radio*my_sin),1)

        px_second = 0
        py_second = 0
        px_minute = 0
        py_minute = 0
        px_hour = 0
        py_hour = 0

        dicmonths= ('NaN','Ene','Feb','Mar','Abr','May','Jun','Jul','Ago','Sep','Oct','Nov','Dic')
        rtc = RTC()
        rtc.datetime((2019, 12, 27, 6, 18, 33, 0, 0)) # set a specific date and time
        while True:
            dt = rtc.datetime() # get date and time
            self.display.fill_rect(0,0,72,12,0)
            self.display.fill_rect(0,12,64,24,0)
            self.display.text("{day:02} de {month}".format(month=dicmonths[dt[1]], day=dt[2]),0,0,1)
            self.display.text("{hour:02}:{minute:02}:{second:02}".format(hour=dt[4], minute=dt[5], second=dt[6]),0,12,1)

            x_hour = math.cos((dt[4]-3 + (dt[5]-15)/60 )*2*math.pi/12)
            y_hour = math.sin((dt[4]-3 + (dt[5]-15)/60 )*2*math.pi/12)
            x_minute = math.cos((dt[5]-15)*2*math.pi/60)
            y_minute = math.sin((dt[5]-15)*2*math.pi/60)
            x_second = math.cos((dt[6]-15)*2*math.pi/60)
            y_second = math.sin((dt[6]-15)*2*math.pi/60)

            self.display.pixel(96+int(radio*0.95*px_second), 32+int(radio*0.95*py_second), 0)
            self.display.pixel(96+int(radio*0.95*x_second), 32+int(radio*0.95*y_second), 1)
            px_second = x_second
            py_second = y_second

            self.display.line(96, 32, 96+int(radio*0.8*px_minute), 32+int(radio*0.8*py_minute), 0)
            self.display.line(96, 32, 96+int(radio*0.8*x_minute), 32+int(radio*0.8*y_minute), 1)
            px_minute = x_minute
            py_minute = y_minute

            self.display.line(96, 32, 96+int(radio*0.5*px_hour), 32+int(radio*0.5*py_hour), 0)
            self.display.fill_rect(96-1+int(radio*0.5*px_hour), 32-1+int(radio*0.5*py_hour), 3, 3, 0)
            self.display.line(96, 32, 96+int(radio*0.5*x_hour), 32+int(radio*0.5*y_hour), 1)
            self.display.fill_rect(96-1+int(radio*0.5*x_hour), 32-1+int(radio*0.5*y_hour), 3, 3, 1)
            px_hour = x_hour
            py_hour = y_hour

            if not rotary_enc['pb'].value():
                time.sleep_ms(500)
                return None

            self.display.show()
            time.sleep_ms(100)


simple_menu/simple_menu.py
import time
import ssd1306
import simple_encoder
from machine import Pin

class simple_menu():
    def __init__(self, display, rotary_enc, menu_obj, settings_obj):
        self.display = display
        self.rotary_enc = rotary_enc
        self.menu_list = menu_obj["menu_list"]
        self.menu_actions = menu_obj["menu_actions"]
        self.settings_obj = settings_obj
        self.em = simple_encoder.simple_encoder (rotary_enc['ra'], rotary_enc['rb'], rotary_enc['pin_irq'])

    def show(self):
        time.sleep_ms(500)
        self.display.fill(0)
        offset_y = 4
        row_offset = 0
        while True:
            row_offset = 0
            if self.em.counter < 0:
                self.em.counter = len(self.menu_list) - 1
            if self.em.counter >= len(self.menu_list):
                self.em.counter = 0
            if self.em.counter > 5 :
                row_offset = self.em.counter - 5
                if self.em.counter >= len(self.menu_list):
                    self.em.counter = 0
            if self.em.counter == 0:
                row_offset = 0
            emc = self.em.counter

            for row in range(row_offset, len(self.menu_list)):
                if emc == row:
                    bg_color = 1
                    fg_color = 0
                    if not self.rotary_enc['pb'].value():
                        time.sleep_ms(500)
                        if self.menu_actions[row] == '<BACK>':
                            return True
                        module = __import__(self.menu_actions[row], globals(), locals(), [self.menu_actions[row]], 0)
                        my_function = getattr(module, self.menu_actions[row])
                        my_function(self.display, self.rotary_enc, self.settings_obj)
                        self.display.fill(0)
                else:
                    bg_color = 0
                    fg_color = 1
                self.display.fill_rect(8, offset_y+(row-row_offset)*10, 120, 10, bg_color)
                self.display.text("{0}".format(self.menu_list[row]), 8, offset_y+(row-row_offset)*10, fg_color)
            self.display.show()
            time.sleep_ms(100)


simple_menu/simple_alphabet.py
import time
import ssd1306
import simple_encoder
from machine import Pin

class simple_alphabet():
    def __init__(self, display, rotary_enc, settings_obj):
        self.display = display
        self.rotary_enc = rotary_enc
        self.alphanumeric_list = list('\b'+'\n'+settings_obj[0]['ALPHABET'])
        self.em = simple_encoder.simple_encoder (rotary_enc['ra'], rotary_enc['rb'], rotary_enc['pin_irq'])
        self.show()

    def show(self):
        self.display.fill(0)
        self.display.text("{0}".format("SELECT LETTER: "), 0, 0, 1)
        self.display.fill_rect(0,36,128,12,1)
        str_push = ""
        self.em.counter = 3
        while True:
            if self.em.counter < 0:
                self.em.counter = len(self.alphanumeric_list) - 1
            if self.em.counter >= len(self.alphanumeric_list):
                self.em.counter = 0

            self.display.fill_rect(0,12,128,12,0)
            actual_char = self.alphanumeric_list[self.em.counter]
            if actual_char == '\b':
                actual_char = '<BACKSPACE>'
            if actual_char == '\n':
                actual_char = '<END>'
            self.display.text("{0}".format(actual_char), 0, 12, 1)

            if not self.rotary_enc["pb"].value():
                if actual_char == '<BACKSPACE>':
                    actual_char = ''
                    if len(str_push) > 0:
                        str_push = str_push[:-1]
                    else:
                        str_push = ''
                if actual_char == '<END>':
                    time.sleep_ms(500)
                    return str_push
                str_push = str_push + str(actual_char)
                self.display.fill_rect(0,36,128,12,1)
                self.display.text("{0:>16}".format(str_push[-16:]), 0, 36, 0)
                # str_to_return = str_push
                time.sleep_ms(500)

            self.display.show()
            time.sleep_ms(100)



simple_menu/main.py
import time
import ssd1306
import simple_menu
from settings import settings
from machine import SPI, I2C, Pin

rotary_enc = {"pb" : Pin(16, Pin.IN), "ra" : Pin(2, Pin.IN), "rb" : Pin(0, Pin.IN), "pin_irq" : Pin.IRQ_FALLING}
st = settings()
settings_obj = st.get_settings()

i2c_obj = I2C(-1, Pin(14), Pin(13))
addr = i2c_obj.scan()
if 0x3C in addr:
    display = ssd1306.SSD1306_I2C(width=128, height=64, i2c=i2c_obj, addr=0x3C, external_vcc=False)
else:
    display = ssd1306.SSD1306_SPI(width=128, height=64, spi=SPI(1), dc=Pin(4), res=Pin(5), cs=Pin(15), external_vcc=False)

menu_list = ['back', 'clock', 'alarm', 'wifi connect', 'wifi server', 'settings', 'reset', 'alphabet']
menu_actions = ['<BACK>', 'clock', 'alarm', 'wificonnect', 'wifiserver', 'settings', 'reset', 'simple_alphabet']
menu_obj = {"menu_list": menu_list, "menu_actions": menu_actions}

menu = simple_menu.simple_menu(display, rotary_enc, menu_obj, settings_obj)

while True:
    display.fill_rect(0,0,128,64,0)
    display.text("{: ^16s}".format("WELCOME"),0,0,1)
    display.text("{: ^16s}".format("PRESS START"),0,30,1)
    display.show()

    if not rotary_enc['pb'].value():
        menu.show()
    time.sleep_ms(100)
